### PR TITLE
add support for environment variables and preopens

### PR DESCRIPTION
**Pull request by dicej**
Per #31, we pass env vars and preopens via `command`, just like CLI arguments.
Later, we plan to add another interface (e.g. `cli-reactor`) for reactors
which need env vars and/or preopens, which will have a single function
`initialize` that takes the same two parameters (but not stdio or CLI arg
parameters).

This also removes unused parts of `wasi-common` (e.g. args, and env vars, and
preopen paths) which have been superceded by the `command` interface.

Our eventual goal is to provide a more explicit interface for environment
variables using WIT star imports such that the guest will import a zero-argument
function (or value, when that is supported) for each variable it needs, allowing
the host to statically verify that it can provide all those variables.  However,
star imports are not yet supported in `wit-bindgen`, and once they are, we'll
likely still need a dynamic interface to support existing software.

Note that I've added a `file_read` test which does not actually do anything yet,
since not all the required host functions have been implemented.  I plan to
address that soon.

### CODE REVIEW

File: host/src/environment.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+#[async_trait::async_trait]
+impl wasi_environment::WasiEnvironment for WasiCtx {
+    async fn get(&mut self) -> anyhow::Result<Vec<(String, String)>> {
+        self.env
+            .iter()
+            .map(|s| {
+                let (k, v) = s.split_at(
+                    s.find('=')
+                        .ok_or_else(|| anyhow::anyhow!("expected to find '=' in {s}"))?,
```
Comment (alexcrichton) [inline host/src/environment.rs:11]: Could this perhaps be stored, at rest, as `Vec<(String, String)>`?

**Review by alexcrichton — APPROVED** [review]
Looks good to me, thanks! I think @sunfishcode will want to look at this to to make sure the idea is ok before merging.

As for the `AlignedBool`, turns out you've uncovered a preexisting issue! I believe the problem is that with a 1-aligned `bool` rustc is placing the flag at the end of the struct, meaning the write at the very end is being out of bounds. The preexisting issue is that while `State` was the size of a page we were actually storing `RefCell<State>` in the page instead of `State`, meaning that `State` actually extended just beyond the page.

The fix I believe is to change the assert to assert on the size of `RefCell<State>` and I had to use `13 * usize` instead of `11` as-is in the code. That I think means the alignment issues go away (and additionally fixes the case where data at the very end of the array could either be out of bounds or corrupt the wasm module just beyond it)

File: host/src/environment.rs (hunk 1 above)
Comment (dicej) [reply host/src/environment.rs:11]: Yeah, I was trying to minimize the changes needed to `wasi-common`, but I agree this is worth doing.

Comment (dicej) [reply host/src/environment.rs:11]: BTW, @pchickey is `StringArray` worth keeping around?  If I switch `WasiCtx::env` to a `Vec<(String, String)>` as @alexcrichton suggests, we'll lose the size checks `StringArray` provides, and I don't know what the original intent of those checks was.  I can also work on generic-izing `StringArray` if that's worthwhile.

Comment (pchickey) [reply host/src/environment.rs:11]: StringArray was really just specific to witx, lets get rid of it!

**Comment by sunfishcode** [comment]
The code here looks reasonable to me. The open question is whether we want to support dynamic environment variables. Previously we were hoping to avoid that, by using star-imports, since dynamic environment variables create implicit contracts between the environment and the component. But since star-imports aren't implemented in wit-bindgen yet, it may make sense to implement this now and revisit this later. I'm talking with other people to see if that makes sense.

**Review by sunfishcode — COMMENTED** [review]
At an initial pass, this looks good.

I see this removes the sched code; #29 significantly revamps the sched code; would be be feasible to reinstate it in this patch, so that #29 can more easily replace it?

**Comment by dicej** [comment]
@sunfishcode I've restored the `sched` code and addressed the formatting issue.  I've also squashed the commits.

Next, I'll work on implementing the host functions needed to make the `file_read` test pass.

**Comment by dicej** [comment]
I've implemented the host `wasi-filesystem` functions `open-at`, `stat`, `seek`, and `close`, so the `file_read` test now passes.

@sunfishcode Anything else you'd like me to change or add here?

**Comment by sunfishcode** [comment]
This looks good!

**Pull request merged**